src/ucp/core/ucp_device.c

```diff
@@ -90,87 +154,281 @@ ucp_device_mem_list_params_check(const ucp_device_mem_list_params_t *params)
+        uct_allocated_memory_t *mem)
 {
+    size_t handle_size = 0;
+    size_t uct_elem_size[UCP_DEVICE_MEM_LIST_MAX_EPS];
+    uint8_t i, j, num_uct_eps;
+    uct_iface_attr_v2_t attr;
     ucs_status_t status;
-    ucp_device_mem_list_handle_t host_handle;
-    uct_allocated_memory_t mem;
-    ucp_memory_info_t mem_info;
-    khiter_t iter;
-    int ret;
+    ucp_worker_iface_t *wiface;
+    ucp_device_mem_list_handle_t handle;
+    uct_mem_h uct_memh;
+    uct_rkey_t uct_rkey;
+    uct_device_mem_element_t *uct_element;
+    const ucp_device_mem_list_elem_t *ucp_element;
+    ucp_md_index_t local_md_index;
+    uint8_t rkey_index;
 
-    status = ucp_device_mem_list_params_check(params);
-    if (status != UCS_OK) {
-        return status;
+    /* For each available lane */
+    for (i = 0;
+         (i < UCP_DEVICE_MEM_LIST_MAX_EPS) && (lanes[i] != UCP_NULL_LANE);
+         i++) {
+        /* Query per transport UCT memory list element size */
+        wiface          = ucp_worker_iface(ep->worker,
+                                           ucp_ep_get_rsc_index(ep, lanes[i]));
+        attr.field_mask = UCT_IFACE_ATTR_FIELD_DEVICE_MEM_ELEMENT_SIZE;
+        status          = uct_iface_query_v2(wiface->iface, &attr);
+        if (status != UCS_OK) {
+            ucs_error("failed to query element size lanes[%u]=%u", i, lanes[i]);
+            return status;
+        }
+
+        ucs_trace("query lane=%u device_mem_element_size=%zu", lanes[i],
+                  attr.device_mem_element_size);
+
+        handle_size     += attr.device_mem_element_size;
+        uct_elem_size[i] = attr.device_mem_element_size;
     }
 
-    /* TODO: Allocate based on lane selection results */
-    status = ucp_mem_do_alloc(ep->worker->context, NULL, sizeof(**handle_p),
-                              UCT_MD_MEM_ACCESS_LOCAL_READ |
-                                      UCT_MD_MEM_ACCESS_LOCAL_WRITE,
-                              UCS_MEMORY_TYPE_CUDA, UCS_SYS_DEVICE_ID_UNKNOWN,
-                              "ucp_device_mem_list_handle_t", &mem);
+    if (i == 0) {
+        ucs_error("failed to select lane");
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    /* Populate handle header */
+    num_uct_eps            = i;
+    handle.version         = UCP_DEVICE_MEM_LIST_VERSION_V1;
+    handle.proto_idx       = 0;
+    handle.num_uct_eps     = num_uct_eps;
+    handle.mem_list_length = params->num_elements;
+    for (i = 0; i < num_uct_eps; i++) {
+        status = uct_ep_get_device_ep(ucp_ep_get_lane(ep, lanes[i]),
+                                      &handle.uct_device_eps[i]);
+        if (status != UCS_OK) {
+            ucs_error("failed to get device_ep for lane=%u", lanes[i]);
+            goto err;
+            ;
+        }
+
+        handle.uct_mem_element_size[i] = uct_elem_size[i];
+    }
+
+    /* Allocate handle on the same device memory */
+    handle_size *= params->num_elements;
+    handle_size += sizeof(handle);
```

Comment:
nextpr: cleanup handle_size calculation, maybe use separate helper var for element_size